### PR TITLE
add sql stuff in init.sc so we don't have to always repeat ourselves

### PR DESCRIPTION
**Pull request by andypetrella**
not more need to `val sqlContext = ...; import implicits...`

cc @vidma @ndjido 

> Note
> only if `reset` is called, sqlContext is a `var` in the `globalScope` object

### CODE REVIEW

**Comment by vidma** [comment]
What if user wants to use a HiveContext instead?
Wouldn't it conflict?

**Comment by ndjido** [comment]
It's really handy. 
Thx!

**Comment by andypetrella** [comment]
@vidma you're right, actually in this case, he probably can redefine a context and import it.
Redefining the one in globalScope would probably better, but it has to be known.

We could also include a metadata to force the usage of the hive context? Is it just a different constructor to use? Or the code is not trivial to foresee?

This is a best effort of course, to get started faster and to avoid 90% of the cases of repitition.

and btw, I hope hive stuff will disappear with time ^^

File: conf/scripts/init.sc [hunk 1]
```diff
@@ -130,16 +134,23 @@ import scala.util.matching.Regex
...
 }
-
-import globalScope.{sparkContext, reset, sc}
+import globalScope.{reset}
 
 reset()
 
+import globalScope.{sparkContext, sc}
+
+@transient val tempSQLC = globalScope.sqlContext
```
Comment (vidma) [inline conf/scripts/init.sc]: shouldn't it  expose the `sqlContext` variable to the user...?

**Comment by vidma** [comment]
well, if I remember it well, it's OK to have multiple sql-contexts (but in `Spark UI` I'd look a bit weird).

---

another related thing, is to think about sharing global settings/imports/deps (in notebook metadata), maybe something like:
```json
{ 
  "inherit_metadata_from": "company_default_metadata.json",
  "customDeps": [
    "com.databricks % spark-csv_2.11 % 1.3.0",
    "company % domain % 0.3-dev"
  ],
}
```
above `customDeps` completely overrides default `customDeps`, all rest is inherited

**Comment by andypetrella** [comment]
Good point @vidma! I was thinking about something similar, like including a notion of priority for each key like for plugins, or maybe the inclusion pattern in hocon 

Also having multiple definition of the custom  environments and a switcher would be lovely

**Comment by vidma** [comment]
> Also having multiple definition of the custom environments and a switcher would be lovely

not bad idea. then the notebook config would contain only the overwrites, and the simplest is to have them to fully replace the default value(s).

File: conf/scripts/init.sc (hunk 1 above)
Comment (andypetrella) [inline conf/scripts/init.sc]: so you think I should rename it as sqlContext right?

**Comment by jarutis** [comment]
Maybe it would be possible to fix `val sqlContext = ...; import implicits...` by allowing any scala code in imports part of metadata instead of a list of packages?

File: conf/scripts/init.sc [hunk 2]
```diff
@@ -130,16 +134,23 @@ import scala.util.matching.Regex
     }
     sparkContext = new SparkContext(conf)
     sparkContext.hadoopConfiguration.set("fs.tachyon.impl", "tachyon.hadoop.TFS")
+
+    sqlContext = new SQLContext(sparkContext)
```
Comment (vidma) [inline conf/scripts/init.sc]: also what if one is using HiveContext instead ?

**Comment by andypetrella** [comment]
@jarutis well... actually... the customImports is not restricting to imports ^^.
The name was reaaaaaaally badly chosen, and it should be renamed.

**Comment by jarutis** [comment]
I remembered, the problem with this approach was that at the moment of evaluation of customImports, we didn't had SparkContext defined (`sc`).

**Comment by andypetrella** [comment]
This will be probably obsolete in spark 2.0.0 and spark session

**Comment by jarutis** [comment]
https://amindinfinite.files.wordpress.com/2014/05/bright-future-ahead-sm.jpg

File: conf/scripts/init.sc (hunk 1 above)
Comment (vidma) [inline conf/scripts/init.sc]: certainly.

**Comment by vidma** [comment]
how about spark-2.0? we have a `SparkSession` there, so it should be a bit different...

**Comment by andypetrella** [comment]
yeah the branch `future-master` will be the one for spark2+ which will allow us to take more actions/decisions regarding this kind of things, including the dependencies we're stuck on due to a mix of spark ones and the server

**Pull request closed without merging**